### PR TITLE
Support text-completion within a `:refer`

### PR DESCRIPTION
**Pull request by anonimitoraf**
Initial shot at implementing support for text-completion for `:refer` blocks

For example:
image

I'm not convinced that my implementation is the cleanest it can be, so feel free to suggest edits and whatnot

### CODE REVIEW

File: src/clojure_lsp/handlers.clj [hunk 1]
```diff
@@ -215,61 +215,73 @@
...
     (when cursor-value
-      (concat
-        (->> local-env
-             (filter (comp :declare :tags))
+      (if (contains? (:tags cursor-usage) :refer)
+        ;; If the cursor is within a :refer then the sole
+        ;; completions should be symbols within the namespace
+        ;; that's being referred
+        (->> (get remotes-by-ns (:ns cursor-usage))
+             (filter #(every? (:tags %) [:declare :public]))
```
Comment (snoe) [inline src/clojure_lsp/handlers.clj:223]: I get a null pointer exception here when `:tags` are nil.

Comment (anonimitoraf) [reply src/clojure_lsp/handlers.clj:223]: Ah, in which cases is `:tags` nil? (So I can add tests for it too)

**Comment by ericdallo** [comment]
Hey @anonimitoraf, I'm not sure when tags are nil, but in this comment of another issue, we can see that there are cases where tags are really nil:
https://github.com/snoe/clojure-lsp/issues/208#issuecomment-749822478

**Comment by anonimitoraf** [comment]
Hey @ericdallo I see. Shall I just filter out the ones with no tags then?

**Comment by ericdallo** [comment]
@anonimitoraf after [this PR](https://github.com/snoe/clojure-lsp/pull/216/files#diff-897a27d95724c2a85947bfa6bc7b5ac1086272aeb125972c1b57749beea71050R209) `tags` doesn't come nil anymore :)
Could you rebase your branch to let me test it again?

**Comment by anonimitoraf** [comment]
@ericdallo Done

Although I'm curious, what's the gist of the solution (for preventing non-nil tags)?

**Comment by ericdallo** [comment]
This was the solution: https://github.com/snoe/clojure-lsp/pull/216/files#diff-897a27d95724c2a85947bfa6bc7b5ac1086272aeb125972c1b57749beea71050R209

File: test/clojure_lsp/handlers_test.clj [hunk 2]
```diff
@@ -375,6 +379,12 @@
               {:label "alpaca/barr" :detail "alpaca.ns" :data "alpaca.ns/barr"}
               {:label "alpaca/bazz" :detail "alpaca.ns" :data "alpaca.ns/bazz"}]
              (handlers/completion "file://b.clj" 3 18))))
+    (testing "complete-within-refering"
+      (reset! db/db db-state)
+      (print (handlers/completion "file://e.clj" 1 38))
```
Comment (ericdallo) [inline test/clojure_lsp/handlers_test.clj:384]: :scissors:

Comment (ericdallo) [reply test/clojure_lsp/handlers_test.clj:384]:
```suggestion
```

**Review by ericdallo — APPROVED** [review]
LGTM
Tested and it works, thanks for your contribution!

**Comment by anonimitoraf** [comment]
Oh whoops, thanks for spotting the print statement and thanks for merging! :)

**Comment by anonimitoraf** [comment]
@ericdallo I get this error in Emacs:

```
Debugger entered--Lisp error: (wrong-type-argument hash-table-p nil)
  lsp--on-diagnostics(#s(lsp--workspace :ewoc nil :server-capabilities #<hash-table equal 14/14 0x1563efaa9c33> :registered-server-capabilities (#s(lsp--registered-capability :id "id" :method "workspace/didChangeWatchedFiles" :options #<hash-table equal 1/1 0x1563f2c48d4b>)) :root "/home/anonimito/work/pegleg/clojure/tung-frontend" :client #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name environment-fn) #<bytecode 0x16466b61b1dc60c>) :test\? #f(compiled-function () #<bytecode 0x1fcae400dcad>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 0/65 0x1563eff439e7> :request-handlers #<hash-table equal 0/65 0x1563f202f7a3> :response-handlers #<hash-table eql 2/65 0x1563f202f7c3> :prefix-function nil :uri-handlers #<hash-table equal 1/65 0x1563f202f783> :action-handlers #<hash-table equal 0/65 0x1563f12ac873> :major-modes (clojure-mode clojurec-mode clojurescript-mode) :activation-fn nil :priority 0 :server-id clojure-lsp :multi-root nil :initialization-options (:dependency-scheme "jar") :custom-capabilities nil :library-folders-fn lsp-clojure--library-folders :before-file-open-fn nil :initialized-fn nil :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil :environment-fn nil :after-open-fn nil :async-request-handlers #<hash-table equal 0/65 0x1563f1630e23> :download-server-fn #f(compiled-function (client callback error-callback update\?) #<bytecode 0x3fb3beaa148bcad>) :download-in-progress? nil :buffers nil) :host-root nil :proc #<process clojure-lsp> :cmd-proc #<process clojure-lsp> :buffers (#<buffer auth.cljs>) :semantic-highlighting-faces nil :semantic-highlighting-modifier-faces nil :extra-client-capabilities nil :status initialized :metadata #<hash-table equal 0/65 0x1563f26f2d49> :watches #<hash-table equal 0/65 0x1563f26f2e6f> :workspace-folders nil :last-id 0 :status-string nil :shutdown-action nil :diagnostics #<hash-table equal 0/65 0x1563f26f2e8f> :work-done-tokens #<hash-table equal 0/65 0x1563f05d6ad7>) nil)
  lsp--on-notification(#s(lsp--workspace :ewoc nil :server-capabilities #<hash-table equal 14/14 0x1563efaa9c33> :registered-server-capabilities (#s(lsp--registered-capability :id "id" :method "workspace/didChangeWatchedFiles" :options #<hash-table equal 1/1 0x1563f2c48d4b>)) :root "/home/anonimito/work/pegleg/clojure/tung-frontend" :client #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name environment-fn) #<bytecode 0x16466b61b1dc60c>) :test\? #f(compiled-function () #<bytecode 0x1fcae400dcad>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 0/65 0x1563eff439e7> :request-handlers #<hash-table equal 0/65 0x1563f202f7a3> :response-handlers #<hash-table eql 2/65 0x1563f202f7c3> :prefix-function nil :uri-handlers #<hash-table equal 1/65 0x1563f202f783> :action-handlers #<hash-table equal 0/65 0x1563f12ac873> :major-modes (clojure-mode clojurec-mode clojurescript-mode) :activation-fn nil :priority 0 :server-id clojure-lsp :multi-root nil :initialization-options (:dependency-scheme "jar") :custom-capabilities nil :library-folders-fn lsp-clojure--library-folders :before-file-open-fn nil :initialized-fn nil :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil :environment-fn nil :after-open-fn nil :async-request-handlers #<hash-table equal 0/65 0x1563f1630e23> :download-server-fn #f(compiled-function (client callback error-callback update\?) #<bytecode 0x3fb3beaa148bcad>) :download-in-progress? nil :buffers nil) :host-root nil :proc #<process clojure-lsp> :cmd-proc #<process clojure-lsp> :buffers (#<buffer auth.cljs>) :semantic-highlighting-faces nil :semantic-highlighting-modifier-faces nil :extra-client-capabilities nil :status initialized :metadata #<hash-table equal 0/65 0x1563f26f2d49> :watches #<hash-table equal 0/65 0x1563f26f2e6f> :workspace-folders nil :last-id 0 :status-string nil :shutdown-action nil :diagnostics #<hash-table equal 0/65 0x1563f26f2e8f> :work-done-tokens #<hash-table equal 0/65 0x1563f05d6ad7>) #<hash-table equal 3/3 0x1563efece0e9>)
  lsp--parser-on-message(#<hash-table equal 3/3 0x1563efece0e9> #s(lsp--workspace :ewoc nil :server-capabilities #<hash-table equal 14/14 0x1563efaa9c33> :registered-server-capabilities (#s(lsp--registered-capability :id "id" :method "workspace/didChangeWatchedFiles" :options #<hash-table equal 1/1 0x1563f2c48d4b>)) :root "/home/anonimito/work/pegleg/clojure/tung-frontend" :client #s(lsp--client :language-id nil :add-on? nil :new-connection (:connect #f(compiled-function (filter sentinel name environment-fn) #<bytecode 0x16466b61b1dc60c>) :test\? #f(compiled-function () #<bytecode 0x1fcae400dcad>)) :ignore-regexps nil :ignore-messages nil :notification-handlers #<hash-table equal 0/65 0x1563eff439e7> :request-handlers #<hash-table equal 0/65 0x1563f202f7a3> :response-handlers #<hash-table eql 2/65 0x1563f202f7c3> :prefix-function nil :uri-handlers #<hash-table equal 1/65 0x1563f202f783> :action-handlers #<hash-table equal 0/65 0x1563f12ac873> :major-modes (clojure-mode clojurec-mode clojurescript-mode) :activation-fn nil :priority 0 :server-id clojure-lsp :multi-root nil :initialization-options (:dependency-scheme "jar") :custom-capabilities nil :library-folders-fn lsp-clojure--library-folders :before-file-open-fn nil :initialized-fn nil :remote? nil :completion-in-comments? nil :path->uri-fn nil :uri->path-fn nil :environment-fn nil :after-open-fn nil :async-request-handlers #<hash-table equal 0/65 0x1563f1630e23> :download-server-fn #f(compiled-function (client callback error-callback update\?) #<bytecode 0x3fb3beaa148bcad>) :download-in-progress? nil :buffers nil) :host-root nil :proc #<process clojure-lsp> :cmd-proc #<process clojure-lsp> :buffers (#<buffer auth.cljs>) :semantic-highlighting-faces nil :semantic-highlighting-modifier-faces nil :extra-client-capabilities nil :status initialized :metadata #<hash-table equal 0/65 0x1563f26f2d49> :watches #<hash-table equal 0/65 0x1563f26f2e6f> :workspace-folders nil :last-id 0 :status-string nil :shutdown-action nil :diagnostics #<hash-table equal 0/65 0x1563f26f2e8f> :work-done-tokens #<hash-table equal 0/65 0x1563f05d6ad7>))
  #f(compiled-function (proc input) #<bytecode -0x17bdb02f5a4ec78d>)(#<process clojure-lsp> "Content-Length: 74\15\n\15\n{\"jsonrpc\":\"2.0\",\"method\":\"t...")
  apply(#f(compiled-function (proc input) #<bytecode -0x17bdb02f5a4ec78d>) (#<process clojure-lsp> "Content-Length: 74\15\n\15\n{\"jsonrpc\":\"2.0\",\"method\":\"t..."))
  explain-pause--wrap-callback(#s(explain-pause-command-record :command process-filter :native nil :parent #s(explain-pause-command-record :command "clojure-lsp" :native nil :parent #s(explain-pause-command-record :command #f(compiled-function () #<bytecode 0x142b2f1623b813c3>) :native nil :parent #s(explain-pause-command-record :command timer :native t :parent #s(explain-pause-command-record :command counsel-recentf :native nil :parent #s(explain-pause-command-record :command root-emacs :native nil :parent nil :executing-time 11917 :entry-snap (24554 330 860966 752000) :too-slow nil :is-profiled nil :under-profile nil :profile nil :depth 0) :executing-time 1695 :entry-snap (24554 324 544512 870000) :too-slow t :is-profiled nil :under-profile nil :profile nil :depth 1) :executing-time 0 :entry-snap nil :too-slow nil :is-profiled nil :under-profile nil :profile nil :depth 2) :executing-time 894 :entry-snap (24554 325 320334 369000) :too-slow t :is-profiled nil :under-profile nil :profile nil :depth 3) :executing-time 0 :entry-snap nil :too-slow nil :is-profiled nil :under-profile nil :profile nil :depth 4) :executing-time 0 :entry-snap nil :too-slow nil :is-profiled nil :under-profile nil :profile nil :depth 5) #f(compiled-function (proc input) #<bytecode -0x17bdb02f5a4ec78d>) #<process clojure-lsp> "Content-Length: 74\15\n\15\n{\"jsonrpc\":\"2.0\",\"method\":\"t...")
  apply(explain-pause--wrap-callback #s(explain-pause-command-record :command process-filter :native nil :parent #s(explain-pause-command-record :command "clojure-lsp" :native nil :parent #s(explain-pause-command-record :command #f(compiled-function () #<bytecode 0x142b2f1623b813c3>) :native nil :parent #s(explain-pause-command-record :command timer :native t :parent #s(explain-pause-command-record :command counsel-recentf :native nil :parent #s(explain-pause-command-record :command root-emacs :native nil :parent nil :executing-time 11917 :entry-snap (24554 330 860966 752000) :too-slow nil :is-profiled nil :under-profile nil :profile nil :depth 0) :executing-time 1695 :entry-snap (24554 324 544512 870000) :too-slow t :is-profiled nil :under-profile nil :profile nil :depth 1) :executing-time 0 :entry-snap nil :too-slow nil :is-profiled nil :under-profile nil :profile nil :depth 2) :executing-time 894 :entry-snap (24554 325 320334 369000) :too-slow t :is-profiled nil :under-profile nil :profile nil :depth 3) :executing-time 0 :entry-snap nil :too-slow nil :is-profiled nil :under-profile nil :profile nil :depth 4) :executing-time 0 :entry-snap nil :too-slow nil :is-profiled nil :under-profile nil :profile nil :depth 5) #f(compiled-function (proc input) #<bytecode -0x17bdb02f5a4ec78d>) (#<process clojure-lsp> "Content-Length: 74\15\n\15\n{\"jsonrpc\":\"2.0\",\"method\":\"t..."))
  #f(compiled-function (&rest callback-args) #<bytecode 0x7b1fd78fb6b4ae3>)(#<process clojure-lsp> "Content-Length: 74\15\n\15\n{\"jsonrpc\":\"2.0\",\"method\":\"t...")
```

Any ideas if any of the recent changes caused this?

**Comment by ericdallo** [comment]
It works for me with `lsp-mode`, make sure you have `lsp-mode` updated

**Comment by anonimitoraf** [comment]
Oh, right. Was there a forwards-compatible change that requires lsp-mode to also be updated?

**Comment by anonimitoraf** [comment]
I'm using doom so, I'll try unpinning lsp-mode to the latest (tomorrow)

**Comment by ericdallo** [comment]
Not sure its related to this pr, you changed only completion and your error is related to lsp diagnostics.

If you need any help you can ask in our discord server: https://discord.gg/swuxy5AAgT

**Comment by anonimitoraf** [comment]
Yea I'm guessing it's other changes (I haven't updated clojure-lsp in a long while). Cool, if the latest version of lsp-mode still doesn't work, I'll ask in the discord, thanks!

**Comment by anonimitoraf** [comment]
Ok, actually that error is just from 1 file but not sure what makes that file special. I'll raise an Issue later.

**Pull request merged**